internal/admin/snwatcher/snwatcher.go

```diff
@@ -168,7 +168,12 @@ func (snw *StorageNodeWatcher) handleHeartbeatTimeout(ctx context.Context, now t
 
 	for snid, ts := range snw.hb {
 		if now.Sub(ts) > snw.tick*time.Duration(snw.heartbeatTimeout) {
-			snw.eventHandler.HandleHeartbeatTimeout(ctx, snid)
+			// TODO: Handle unresponsive heartbeat concurrently.
+			func(ctx context.Context, snid types.StorageNodeID) {
+				ctx, cancel := context.WithTimeout(ctx, snw.failureHandlerTimeout)
```

Comment:
If we have more than 3 heartbeat failed nodes, adjusting the timeout alone will not solve it. As you said, you have to handle it concurrent.

Reply:
Handling heartbeat timeout concurrently will be a next issue.
If there is no timeout for handling failed nodes, the loop to check and handle heartbeat timeout is stuck.